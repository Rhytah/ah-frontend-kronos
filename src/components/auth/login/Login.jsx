import React from 'react';
import PropTypes from 'prop-types';
import '../../../assets/scss/Login.scss';
import '../../../assets/scss/SocialLogin.scss';
import Loading from '../../common/Loading';

import FacebookOption from '../../social/facebook';
import GoogleOption from '../../social/google';

export default function Login(props) {
  const { handleSubmit, onInputChange, isLoading } = props;
  return (
    <div className="login-clean">
      <form onSubmit={handleSubmit}>
        <h2 className="sr-only">Login Form</h2>
        <div className="illustration">
          <i className="icon ion-ios-navigate" />
        </div>
        <div className="social-apps" />
        <div className="text-center social-btn">
          <FacebookOption />
          <br />
          <GoogleOption />
          <div className="or-seperator">
            <span className="or">- OR - </span>
          </div>
        </div>


        {isLoading
          ? <Loading />
          : (
            <div>
              <div className="form-group">
                <input id="email" type="email" name="email" onChange={onInputChange} placeholder="Email" className="form-control" />
              </div>
              <div className="form-group">
                <input
                  id="password"
                  type="password"
                  name="password"
                  onChange={onInputChange}
                  placeholder="Password"
                  className="form-control"
                />
              </div>
              <div className="form-group">
                <button className="btn btn-primary btn-block" type="submit">
                  Login
                </button>
              </div>
            </div>
          )}
        <a href="www.google.com" className="forgot">
          Forgot your email or password?
        </a>
      </form>
    </div>
  );
}


Login.propTypes = {
  onInputChange: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
};

Login.defaultProps = {
  isLoading: false,
};
