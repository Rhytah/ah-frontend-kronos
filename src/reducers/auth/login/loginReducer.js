import { LoginConstants } from '../../../actions/actionTypes';


const loginReducer = (state = {
  isUserLoggedIn: false,
  token: '',
  user: {},
  username: '',
  logged_in: false,
}, { type, payload }) => {
  switch (type) {
    case LoginConstants.LOGIN_USER_SUCCESS:
      return {
        ...state,
        user: payload.user,
        isUserLoggedIn: true,
        logged_in: true,
      };
    case LoginConstants.LOGIN_USER_FAILED:
      return {
        ...state,
        errors: payload.errors.error[0],
        isUserLoggedIn: false,
      };
    default:
      return state;
  }
};

export default loginReducer;
