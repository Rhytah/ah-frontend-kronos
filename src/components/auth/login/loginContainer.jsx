import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';
import Login from './Login';
import loginUserAction from '../../../actions/auth/loginAction';
import { toastFailure } from '../../../utils/toast';

export class LoginContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      isLoading: false,
    };
  }

  onInputChange = (e) => {
    e.preventDefault();
    this.setState({
      [e.target.name]: e.target.value,
    });
  }

  handleSubmit = async (e) => {
    this.setState({ isLoading: true });
    e.preventDefault();
    const { email, password } = this.state;
    const loginInfo = {
      email,
      password,
    };
    const { loginAction } = this.props;

    if (this.validateEmailPasswordEntry(email, password)) {
      return;
    }

    await loginAction(loginInfo);

    this.setState({ isLoading: false });
    this.handleClearForm();
  }

  validateEmailPasswordEntry(email, password) {
    if ((email === '') || (password === '')) {
      toast.dismiss();
      toastFailure('Both Email & Password Details Required', 'A');
      this.setState({ isLoading: false });
      return true;
    }
    return false;
  }

  handleClearForm() {
    this.setState({
      email: '',
      password: '',
    });
  }

  render() {
    const { isLoading } = this.state;
    return (
      <Login
        handleSubmit={this.handleSubmit}
        onInputChange={this.onInputChange}
        isLoading={isLoading}
      />
    );
  }
}

LoginContainer.propTypes = {
  loginAction: PropTypes.func.isRequired,
};


export const mapStateToProps = state => ({
  loginReducer: state.loginReducer,
});

export default connect(mapStateToProps, { loginAction: loginUserAction })(LoginContainer);
