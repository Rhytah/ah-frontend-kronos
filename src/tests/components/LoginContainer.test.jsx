import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { toast } from 'react-toastify';
import { mapStateToProps, LoginContainer } from '../../components/auth/login/loginContainer';

describe('loginContainer Component', () => {
  let wrapper;
  const props = {
    loginAction: jest.fn(),
    isUserLoggedIn: true,
    user: { auth_token: 'fjjfj', username: 'sjjs', email: 'jdjdj' },
    e: { preventDefault: jest.fn() },
  };

  beforeEach(() => {
    wrapper = shallow(<LoginContainer {...props} />);
    jest.resetAllMocks();
  });

  it('matches snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should set state when onchangeInput is called', () => {
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'email',
        value: 'valuer',
      },
    };
    const instance = wrapper.instance();
    instance.onInputChange(event);
    expect(instance.state.email).toBe('valuer');
  });

  it('should call handleSubmit function', async () => {
    const instance = wrapper.instance();
    instance.setState({
      email: 'manorldsapiens@gmail.com',
      password: 'Jakejake123',
    });
    const event = {
      preventDefault: jest.fn(),
    };

    jest.spyOn(instance, 'handleClearForm');
    await instance.handleSubmit(event);
    expect(props.loginAction).toHaveBeenCalled();
    expect(instance.handleClearForm).toHaveBeenCalled();
  });

  it('should not call loginAction() function without data', async () => {
    const instance = wrapper.instance();
    instance.setState({
      email: '',
      password: '',
    });
    const event = {
      preventDefault: jest.fn(),
    };
    jest.spyOn(instance, 'handleClearForm');
    await instance.handleSubmit(event);
    expect(props.loginAction).not.toHaveBeenCalled();
    expect(instance.handleClearForm).not.toHaveBeenCalled();
  });


  it('should clear local state when form is submitted', () => {
    const event = {
      setState: jest.fn(),
    };
    wrapper.instance().handleClearForm(event);
    expect(wrapper.instance().state.email).toBe('');
  });

  it('should validate when no password and email are entered into the form', () => {
    const instance = wrapper.instance();

    instance.setState({ email: '', password: '', isLoading: true });
    jest.spyOn(toast, 'error');
    instance.validateEmailPasswordEntry('', '');
    expect(instance.state.isLoading).toBe(false);
  });

  it('should validate when correct password and email are entered into the form', () => {
    const instance = wrapper.instance();
    instance.setState({ email: 'manorldsapiens@gmail.com', password: 'Jakejake123', isLoading: true });
    jest.spyOn(toast, 'error');
    instance.validateEmailPasswordEntry('manorldsapiens@gmail.com', 'Jakejake123');
    expect(instance.state.isLoading).toBe(true);
  });

  it('Testing mapStateToProps', () => {
    const initialState = {
      loginReducer: {},
    };
    expect(mapStateToProps(initialState)).toEqual({
      loginReducer: {},
    });
  });
});
