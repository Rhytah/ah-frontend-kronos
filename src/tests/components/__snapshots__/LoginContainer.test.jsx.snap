// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginContainer Component matches snapshot 1`] = `
<Login
  handleSubmit={[Function]}
  isLoading={false}
  onInputChange={[Function]}
/>
`;
